src/main/java/net/oauth/jsontoken/AsyncJsonTokenParser.java

```diff
@@ -143,4 +171,35 @@ public AsyncJsonTokenParser(
     return Futures.transformAsync(futureVerifiers, checkNullFunction, executor);
   }
 
+  /**
+   * Converts exceptions (when applicable) to {@link InvalidJsonTokenException}
+   * for better exception handling in the asynchronous parser.
+   *
+   * @return a failed {@link ListenableFuture}
+   */
+  private ListenableFuture getFailedFuture(Exception e) {
+    Throwable cause = e.getCause();
+    if (cause != null && InvalidJsonTokenException.class.isInstance(cause)) {
+      InvalidJsonTokenException invalidJsonTokenException = (InvalidJsonTokenException) cause;
+      if (invalidJsonTokenException.getErrorCode().equals(ErrorCode.ILLEGAL_STATE)) {
```

Comment:
Why do we still have to do this?

Reply:
We specifically added an Illegal State error code to be able to deal with thrown SignatureException where IllegalStateException should have been thrown. Specifically, here: https://github.com/google/jsontoken/blob/master/src/main/java/net/oauth/jsontoken/JsonToken.java#L341